From 643c021cb6bfd13d1d2d140d4d05cb2040707f69 Mon Sep 17 00:00:00 2001
From: kelson8 <kelson@kelsoncraft.net>
Date: Tue, 15 Sep 2026 13:22:18 -0400
Subject: [PATCH 159/161] feat: Add stubs for setting the zone info in lua

* I have mostly replicated the original scripts for this, although it needs tested and still needs some work.

* Cleanup code in lua_test a bit.

* Label some more items in the code.

* Update changelog, change version to 1.2.14-4a.
---
 Changelog.md                       |  14 +-
 src/core/FileLoader.cpp            |   2 +
 src/extras/lua/lua_garages.cpp     |   5 +
 src/extras/lua/lua_test.cpp        | 340 ++++-------------------------
 src/extras/lua/lua_world.cpp       | 257 ++++++++++++++++++++--
 src/extras/lua/lua_world.h         |   6 +
 src/extras/menus/misc_menu.cpp     |   1 +
 src/extras/utils/defines.cpp       |   2 +-
 src/extras/utils/freeroam_util.cpp |  61 +++++-
 src/save/GenericGameStorage.cpp    |  67 +++---
 10 files changed, 401 insertions(+), 354 deletions(-)

diff --git a/Changelog.md b/Changelog.md
index 0504390c..7117c0fe 100644
--- a/Changelog.md
+++ b/Changelog.md
@@ -923,7 +923,7 @@ feat: Add get game hour and minute to lua
 
 * Update changelog.
 
-### Latest 1.2.14-3a
+### 1.2.14-3a
 
 **Commit 1**
 
@@ -980,3 +980,15 @@ feat: Move version stuff into header
 * Add 'SAVE_FILE_VERSION' and 'SAVE_FILE_FORMAT' for use in lua.
 
 * Update changelog.
+
+### Latest 1.2.14-4a
+
+feat: Add stubs for setting the zone info in lua
+
+* I have mostly replicated the original scripts for this, although it needs tested and still needs some work.
+
+* Cleanup code in lua_test a bit.
+
+* Label some more items in the code.
+
+* Update changelog, change version to 1.2.14-4a.
diff --git a/src/core/FileLoader.cpp b/src/core/FileLoader.cpp
index dbf2cad9..b7000962 100644
--- a/src/core/FileLoader.cpp
+++ b/src/core/FileLoader.cpp
@@ -64,6 +64,8 @@ CFileLoader::LoadLevel(const char *filename)
 	fd = CFileMgr::OpenFile(filename, "r");
 	assert(fd > 0);
 
+	// TODO Look into using this, although some of these files are just basic text files below.
+	// It would still be useful to make a reader in python or something for the game data.
 	for(line = LoadLine(fd); line; line = LoadLine(fd)){
 		if(*line == '#')
 			continue;
diff --git a/src/extras/lua/lua_garages.cpp b/src/extras/lua/lua_garages.cpp
index 0154aa3f..150a1feb 100644
--- a/src/extras/lua/lua_garages.cpp
+++ b/src/extras/lua/lua_garages.cpp
@@ -37,6 +37,11 @@ extern "C"
 		lua_setfield(L, -2, "set");
 
         // TODO Fix these below to work, I'm not sure of the garage id for these values.
+        // TODO Return the garage ID for these values.
+        // I had to do the same concept with the blips.
+        // Although I couldn't figure out how to return the blip ID without creating a new one.
+        // There has to be a way to store the current blips in C++, to be obtained from lua.
+        // I would like to do this without creating more then one blip in an area and make it easily modifiable.
 
         // Open garage
         lua_pushcfunction(L, openGarageLua);
diff --git a/src/extras/lua/lua_test.cpp b/src/extras/lua/lua_test.cpp
index 826918a5..198e5395 100644
--- a/src/extras/lua/lua_test.cpp
+++ b/src/extras/lua/lua_test.cpp
@@ -540,49 +540,7 @@ LuaTest::RemoveFinishedThreads()
 
 //------- Begin cheat functions --------//
 
-// Test, this may or may not work.
-void
-GetBlipIcon()
-{
-	CPlayerPed *pPed = FindPlayerPed();
-	CVector pPlayerPos = FindPlayerCoors();
-
-	CRadar radar = CRadar();
-
-	// I wonder how this one works.
-	// int blip = radar.SetCoordBlip(eBlipType::BLIP_CAR, pPlayerPos, RADAR_TRACE_RED, BLIP_DISPLAY_BOTH);
-	// radar.SetBlipSprite();
-}
-
-// TODO Figure out how to do this, and where to place this for the cheats to work.
-// Got most of this from CPad::AddToPCCheatString
-#ifdef LUA_CHEAT_TEST
-
-char KeyBoardCheatString[30];
-
-extern "C" int
-CheatTest(lua_State *L)
-{
-	char c;
-	for(int32 i = ARRAY_SIZE(KeyBoardCheatString) - 2; i >= 0; i--) KeyBoardCheatString[i + 1] = KeyBoardCheatString[i];
-
-	KeyBoardCheatString[0] = c;
-
-	// These can be generated by reversing the letters to it.
-	// So if you want a 'KILLME' cheat, it would be 'EMLLIK'
-#define _CHEATCMP(str) strncmp(str, KeyBoardCheatString, sizeof(str) - 1)
-#ifdef LUA_TEST
-	// BLOWMEUP
-	// if(!_CHEATCMP("PUEMWOLB"))
-	//	// Run the init lua function, which can test the cheats
-#endif // LUA_TEST
-
-#undef _CHEATCMP
-}
-
-#endif // LUA_CHEAT_TEST
-
-#ifdef _IMGUI_TEST
+#if 0
 extern "C" int
 toggleImGuiLua(lua_State *L)
 {
@@ -590,84 +548,7 @@ toggleImGuiLua(lua_State *L)
 	imGuiFunctions.ImGuiActive = !imGuiFunctions.ImGuiActive;
 	return 1;
 }
-#endif
-
-/**
- * This is a basic cheat code test for lua, kind of pointless but here anyways.
- * 
- * Somewhat replicates the games cheat codes into lua for easier use.
- * 
- * Usage in lua:
- * 
- * -- This should blow up the player
- * game.cheat("BLOWMEUP")
- * 
- * -- This will kill the player
- * game.cheat("KILLME")
- */
-// extern "C" int
-// CheatTestNew(lua_State *L)
-// {
-// 	// Begin new test
-// 	// char KeyBoardCheatString[30];
-// 	// char c;
-// 	// for(int32 i = ARRAY_SIZE(KeyBoardCheatString) - 2; i >= 0; i--) KeyBoardCheatString[i + 1] = KeyBoardCheatString[i];
-
-// 	// KeyBoardCheatString[0] = c;
-
-// 	// These can be generated by reversing the letters to it.
-// 	// So if you want a 'KILLME' cheat, it would be 'EMLLIK'
-// 	// #define _CHEATCMP(str) strncmp(str, KeyBoardCheatString, sizeof(str) - 1)
-
-// 	// End new test
-
-// 	CPlayerPed *pPed = FindPlayerPed();
-// 	VehicleFunctions vehicleFunctions = VehicleFunctions();
-
-// #ifdef _IMGUI_TEST
-// 	ImGuiFunctions imGuiFunctions = ImGuiFunctions();
-// #endif
-
-// 	std::string cheatString = lua_tostring(L, 1);
-// 	// const char* cheatString = lua_tostring(L, 1);
-
-// 	if(cheatString == "BLOWMEUP") {
-// 		// Check if player is in vehicle, blow them up.
-// 		if(!pPed->EnteringCar() && pPed->bInVehicle) {
-// 			CVehicle *currentVeh = pPed->m_pMyVehicle;
-// 			currentVeh->m_fHealth = 0;
-// 		}
-
-// 	} else if(cheatString == "KILLME") {
-// 		//} else if(!_CHEATCMP("KILLME")) {
-// 		// Kill the player, print the message "You have been exterminated"
-// 		FindPlayerPed()->InflictDamage(nil, WEAPONTYPE_UNARMED, 1000.0f, PEDPIECE_TORSO, 0);
-// 		FindPlayerPed()->m_fArmour = 0;
-// 		LogFunctions::LogInfo(Defines::luaLogFile, "Player has been killed");
-// 	} else if(cheatString == "SPAWNCAR") {
-// 		// This works!! I didn't think spawning a car would work.
-// 		if(lua_isnumber(L, 2)) {
-// 			int vehicleModel = lua_tointeger(L, 2);
-// 			vehicleFunctions.SpawnVehicle(vehicleModel, false, false);
-
-// 			// Test for toggling the ImGui menu
-// 		}
-// #ifdef _IMGUI_TEST
-// 		else if(cheatString == "IMGUITG") {
-// 			// ImGuiFunctions::ImGuiDone = !ImGuiFunctions::ImGuiDone;
-// 			// imGuiFunctions.ImGuiDone = !imGuiFunctions.ImGuiDone;
-// 			TRACE("Toggled");
-// 			// toggleImGuiLua();
-// 		}
-
-// #endif
-// 	}
-// 	// switch(cheatString) {
-// 	//
-// 	// }
-
-// 	return 1;
-// }
+#endif //
 
 //------- End cheat functions --------//
 
@@ -726,12 +607,13 @@ LogInfoMsg() {
 	// LogFunctions::LogInfo(Defines::logFile, message);
 }
 
-/// <summary>
-/// I had to add this to the class to fix it.
-/// I got this working, I had to add lua_State *L and get the lua state
-/// To move this back into the init function if needed:
-/// Rename the 'L' variable in each lua_register to 'm_luaVM'
-/// </summary>
+/**
+ * I had to add this to the class to fix it.
+ * I got this working, I had to add lua_State *L and get the lua state.
+ * 
+ * To move this back into the init function if needed:
+ * Rename the 'L' variable in each lua_register to 'm_luaVM'
+ */
 void
 LuaTest::RegisterLuaFunctions(lua_State *L)
 {
@@ -763,9 +645,11 @@ LuaTest::RegisterLuaFunctions(lua_State *L)
 
 	// Toggle for ImGui
 	// TODO Remove this, I don't think this is needed here.
-#ifdef _IMGUI_TEST
+	// I have disabled this, it's not in use
+// #ifdef _IMGUI_TEST
+#if 0
 	lua_register(L, "toggle_imgui", toggleImGuiLua);
-#endif // _IMGUI_TEST
+#endif //
 
 	//----
 	// I will need to rewrite some of my examples but once I fix this,
@@ -917,9 +801,10 @@ For keybinds:
 */
 
 // clang-format off
-/// <summary>
-/// Run the events when the player is busted
-/// </summary>
+
+/**
+ * Run the events when the player is busted
+ */
 void
 LuaTest::BustedLuaEvents()
 {
@@ -927,9 +812,9 @@ LuaTest::BustedLuaEvents()
 		return;
 }
 
-/// <summary>
-/// Run the events when the player is wasted
-/// </summary>
+/**
+ * Run the events when the player is wasted
+ */
 void
 LuaTest::WastedLuaEvents()
 {
@@ -937,9 +822,9 @@ LuaTest::WastedLuaEvents()
 		return;
 }
 
-/// <summary>
-/// Run the events when the keybind is pressed
-/// </summary>
+/**
+ * Run the events when the keybind is pressed
+ */
 void
 LuaTest::LuaKeybindEvents()
 {
@@ -1023,12 +908,12 @@ LuaTest::LuaKeybindEvents()
 
 }
 
-
-// TODO Set this up to run, for now it is just not hooked up to anything.
-
-/// <summary>
-/// Run the events on game init
-/// </summary>
+/**
+ * Run the events on game init, not in use.
+ * 
+ * TODO Why is this here? Possibly remove this later I don't think I'll use it..
+ * I guess it was a leftover from something I was going to do before.
+ */
 void
 LuaTest::SetupLua()
 {
@@ -1038,27 +923,6 @@ LuaTest::SetupLua()
 	}
 }
 
-
-
-/// <summary>
-/// TODO Fix this to work, I would like to stop using the global namespace in Lua.
-/// </summary>
-/// <param name="L"></param>
-/// <returns></returns>
-extern "C" {
-	int LuaOpenTestFunctions(lua_State* L) {
-		static const luaL_Reg mylib[] = {
-			{"create_vehicle_test", createVehicleLua},
-			{NULL, NULL}                       // Sentinel to mark the end of the array
-		};
-
-		luaL_newlib(L, mylib);
-
-		return 1;
-	}
-}
-
-
 /**
  * Setup the freeroam script for init, this is only called if DISABLE_GAME_SCRIPTS is toggled.
  */
@@ -1319,23 +1183,22 @@ LuaTest::SetupFreeroamScript()
 		}
 
 		lua_pop(m_luaVM, 1);
-
-		// gbFadeOnDeath
-
 	}
 #endif // DISABLE_GAME_SCRIPTS - Disable free roam script with main mission scripts.
 }
 
-/// <summary>
-/// Taken idea from this in CLuaMain.cpp within MTA SA.
-/// https://github.com/multitheftauto/mtasa-blue/blob/master/Client/mods/deathmatch/logic/lua/CLuaMain.cpp
-/// TODO Separate the functions, I should be able to separate the init from the functions
-/// </summary>
+
+/**
+ * Init the lua system, this isn't called if DISABLE_GAME_SCRIPTS is active.
+ * For normal game flow, SetupFreeroamScript is for my custom lua mission script replacement.
+ * 
+ * Taken idea from this in CLuaMain.cpp within MTA SA.
+ * https://github.com/multitheftauto/mtasa-blue/blob/master/Client/mods/deathmatch/logic/lua/CLuaMain.cpp
+ */
 void
 LuaTest::InitLua()
 {
 	Defines defines = Defines();
-	// LuaTest luaTest = LuaTest();
 
 	// This requires my custom TcpClient and OpensslTest files from the internal Git repos.
 	// This does send a message to the TcpServer from here.
@@ -1358,38 +1221,6 @@ LuaTest::InitLua()
 	// TRACE("KCNet - Lua Test started");
 	//
 
-	// Set the instruction count hook
-	// TODO Fix this
-	// lua_sethook(m_luaVM, InstructionCountHook, LUA_MASKCOUNT, HOOK_INSTRUCTION_COUNT);
-
-	// Test moving into here
-	// Well this crashes it like this
-	//luaTest.RegisterLuaFunctions();
-
-	//lua_State *L = LuaTest::GetInstance().GetLuaState();
-
-	// This seems to work now.
-	// lua_register(m_luaVM, "set_position", SetPlayerPositionLua);
-	// Well like this it works in the imgui_commands but not on init, I cannot press the keybind for it.. Oh wait..
-	// I got this working as a command in imgui_commands
-	// Usage: set_position(posX, posY, posZ)
-	//lua_register(L, "set_position", SetPlayerPositionLua);
-
-
-	// I finally got this working, I had to pass the lua state to it!
-
-	// Moved into the constructor
-	// TODO Migrate other functions to using this in the constructor.
-	// Reverted
-	//luaTest.RegisterLuaFunctions(m_luaVM);
-
-	// TODO Make lua use this instead of cluttering up the main namespace:
-	//luaL_requiref(m_luaVM, "testfunctions", LuaOpenTestFunctions, 1);
- //   // Pop the loaded module table from the stack
- //   lua_pop(m_luaVM, 1);
-
-	//
-
 	// TODO Hook this up to the ImGui command test.
 
 	//----------------
@@ -1410,6 +1241,7 @@ LuaTest::InitLua()
 
 	// I got this working, it now runs some changes to the C++ code, mostly just respawns.
 	// Toggle never wanted, toggle infinite health, toggle vehicles, and toggle peds.
+	// This file runs if DISABLE_GAME_SCRIPTS is not enabled in config.h.
 
 	// TODO Rename this in the future.
 	if(CheckLua(m_luaVM, luaL_dofile(m_luaVM, defines.luaScript.c_str()))) {
@@ -1472,57 +1304,11 @@ LuaTest::InitLua()
 		//	LogFunctions::LogError(defines.logFile, "The lua file " + defines.luaLocationsScript + "Couldn't be loaded!");
 		//}
 
-			// TODO Test this
-			// Pop the table from the stack
-			//lua_pop(m_luaVM, 1);
-			//
-
-		// This works, I can get booleans from Lua now!
-	//	lua_getglobal(m_luaVM, "test_bool_file2");
-	//	if(lua_isboolean(m_luaVM, -1)) {
-	//		bool test_bool_file2 = static_cast<float>(lua_toboolean(m_luaVM, -1));
-	//		// LuaTest::toggleRespawnPoint = toggleRespawnPoint;
-
-	//		// I've never really used this much before.
-	//		std::string toggleTestFile2 =
-	//		    test_bool_file2 ? "Boolean enabled in " + defines.luaLocationsScript : "Boolean disabled in " + defines.luaLocationsScript;
-	//		// Disabled, this is spammy.
-	//		//LogFunctions::LogInfo(defines.logFile, toggleTestFile2);
-	//		//
-	//		// std::string toggleRespawnString = toggleRespawnPoint ? "Respawn point enabled" : "Respawn point disabled";
-	//	}
-	//}
-
-
 	//-------------------
 	// First lua file
 	//-------------------
 
-	// lua_getglobal(m_luaVM, "create_player");
 	// https://stackoverflow.com/questions/76605585/calling-lua-function-from-c-minimal-example-results-in-lua-errrun#76607016
-	// TODO Test this, should get the create_player function on init.
-	// int result;
-	// result = lua_pcall(m_luaVM, 0, 0, 0);
-
-	// if (result != LUA_OK) {
-    // 	fprintf(stderr, "Error running lua: %s\n", lua_tostring(m_luaVM, -1));
-	// }
-
-	// TODO Figure out how to get this.
-	// For now I'll just use the toggles above.
-	//lua_getglobal(m_luaVM, "lose_weapons_bool");
-	//if (lua_isboolean(m_luaVM, -1)) 
-	//{ 
-	//	bool toggleLoseWeaponsOnDeath = static_cast<bool>(lua_toboolean(m_luaVM, -1));
-	//	LuaTest::toggleLoseWeaponsOnDeath = toggleLoseWeaponsOnDeath;
-
-	//	std::string toggleWeaponsDeathString = luaTest.toggleLoseWeaponsOnDeath 
-	//		? "You will no longer lose weapons on death" 
-	//		: "You will now lose weapons on death";
-
-	//	 LogFunctions::LogInfo(defines.luaLogFile, toggleWeaponsDeathString);
-
-	//}
 
 	// Test for setting the time pass variable when wasted or busted, this is what changes the time.
 	lua_getglobal(m_luaVM, "time_to_pass");
@@ -1549,8 +1335,6 @@ LuaTest::InitLua()
 		// LogFunctions::LogInfo(defines.luaLogFile, toggleRespawnString);
 	}
 
-	// if(lua_isinteger(m_luaVM, -1))
-
 	// Run a function in lua
 
 	// Disabled this for now, it will be useful for later.
@@ -1615,53 +1399,5 @@ lua_setglobal(m_luaVM, "localPlayer");
 */
 }
 
-// TODO Fix this, possibly use the same concept from MTA SA, if I can get it to work.
-#ifdef LUA_TEST
-
-// void
-// HelloWorldTest(lua_State *L)
-//{
-//
-//	// Lua states is needed, to do anything in lua
-//	// TODO Figure this out in a function other then main
-//	//lua_State *L = luaL_newstate();
-//
-//	// Open the libraries, without this, stuff like math won't work
-//	//luaL_openlibs(L);
-//
-//
-//	// Make this run from a file
-//	if(CheckLua(L, luaL_dofile(L, "src/scripts/test.lua"))) {
-//
-//		// Run a function in lua
-//		lua_getglobal(L, "AddStuff");
-//		if(lua_isfunction(L, -1)) {
-//			lua_pushnumber(L, 3.5f);
-//			lua_pushnumber(L, 7.2f);
-//
-//			// The pcall function executes the function.
-//			if(CheckLua(L, lua_pcall(L, 2, 1, 0))) {
-//				float number1 = (float)lua_tonumber(L, -1);
-//				std::cout << "[C++] Called in Lua 'AddStuff(3.5f, 7.2f)', got " << number1 << std::endl;
-//			}
-//		}
-//
-//		// Get the program name
-//		// lua_getglobal(L, "ProgamName");
-//		// if (lua_isstring(L, -1))
-//		//{
-//		//	std::string programName = (std::string)lua_tostring(L, -1);
-//		//	std::cout << "Program name: " << programName << std::endl;
-//		//}
-//
-//	}
-//
-//
-//
-//	// Cleanup the lua vm
-//	lua_close(L);
-// }
-
-#endif // LUA_TEST
 
 #endif // LUA_TEST
\ No newline at end of file
diff --git a/src/extras/lua/lua_world.cpp b/src/extras/lua/lua_world.cpp
index b72bb3e7..17ef9b64 100644
--- a/src/extras/lua/lua_world.cpp
+++ b/src/extras/lua/lua_world.cpp
@@ -162,6 +162,22 @@ extern "C"
 		lua_pushcfunction(L, switchCarGeneratorLua);
 		lua_setfield(L, -2, "switch_car_generator");
 
+		// Zones
+		// TODO Fix these, a few of these are implemented but untested.
+		lua_pushcfunction(L, setZonePedInfoLua);
+		lua_setfield(L, -2, "set_zone_ped_info");
+
+		lua_pushcfunction(L, setZoneCarInfoLua);
+		lua_setfield(L, -2, "set_zone_car_info");
+
+		lua_pushcfunction(L, setZoneCivilianCarInfoLua);
+		lua_setfield(L, -2, "set_zone_civilian_car_info");
+
+		lua_pushcfunction(L, setZoneGroupLua);
+		lua_setfield(L, -2, "set_zone_group");
+		//
+
+
 		// World namespace
     	lua_setglobal(L, "world");
 	}
@@ -326,7 +342,7 @@ extern "C"
 	}
 
 	/**
-	 * Switch the roads off.
+	 * Switch the vehicle roads off.
 	 * 
 	 * Usage (Takes two vectors for a leftBottom, and rightTop values):
 	 * https://library.sannybuilder.com/#/vc/script/extensions/default/01E8
@@ -347,23 +363,37 @@ extern "C"
 		// If this works, then I'll switch the others to using this format also.
 		// Well this didn't work..
 		// I'm not even sure how this function works honestly, I might just make another one that does this.
+		ThePaths.SwitchRoadsOffInArea(
+			leftBottom->x, 
+			rightTop->x,
+			leftBottom->y,
+			rightTop->y,
+			leftBottom->z, 
+			rightTop->z,
+			true);
+
 		// ThePaths.SwitchRoadsOffInArea(
 		// 	leftBottom->x, 
-		// 	rightTop->x,
-		// 	leftBottom->y,
-		// 	rightTop->y,
-		// 	leftBottom->z, 
-		// 	rightTop->z,
+		// 	leftBottom->y, 
+		// 	leftBottom->z,			
+		// 	rightTop->x, 
+		// 	rightTop->y, 
+		// 	rightTop->z, 
 		// 	true);
 
-		ThePaths.SwitchRoadsOffInArea(
-			leftBottom->x, 
-			leftBottom->y, 
-			leftBottom->z,			
-			rightTop->x, 
-			rightTop->y, 
-			rightTop->z, 
-			true);
+		// For debugging
+		// Well this displays a message, why doesn't it turn off the paths for these roads?
+		std::string roadsMsg = "Vehicle roads have been switched off for: "
+		+ std::string("leftBottomX: ") + std::to_string(leftBottom->x) 
+		+ " leftBottomY: " + std::to_string(leftBottom->y)
+		+ " leftBottomZ: " + std::to_string(leftBottom->z)
+		
+		+ " rightTopX: " + std::to_string(rightTop->x)
+		+ " rightTopY: " + std::to_string(rightTop->y)
+		+ " rightTopZ: " + std::to_string(rightTop->z);
+
+		// LogFunctions::LogInfo(Defines::luaLogFile, roadsMsg);
+		//
 
 
 		// SwitchRoadsOffInArea
@@ -371,7 +401,7 @@ extern "C"
 	}
 
 	/**
-	 * Switch the roads on.
+	 * Switch the vehicle roads on.
 	 * 
 	 * Usage (Takes two vectors for a leftBottom, and rightTop values):
 	 * https://library.sannybuilder.com/#/vc/script/extensions/default/01E7
@@ -387,14 +417,15 @@ extern "C"
     		return luaL_error(L, "world.switch_roads_on expects two vector tables with x, y and z values." );
 		}
 
-		ThePaths.SwitchRoadsOffInArea(			leftBottom->x, 
+		ThePaths.SwitchRoadsOffInArea(leftBottom->x, 
 			leftBottom->y, 
 			leftBottom->z,			
 			rightTop->x, 
 			rightTop->y, 
 			rightTop->z, 
 			false);
-		return 1;
+
+		return 0;
 	}
 
 	/**
@@ -413,6 +444,15 @@ extern "C"
     		return luaL_error(L, "world.switch_ped_roads_off expects two vector tables with x, y and z values." );
 		}
 
+		// ThePaths.SwitchPedRoadsOffInArea(
+		// 	leftBottom->x, 
+		// 	rightTop->x,
+		// 	leftBottom->y,
+		// 	rightTop->y,
+		// 	leftBottom->z, 
+		// 	rightTop->z,
+		// 	true);
+
 		ThePaths.SwitchPedRoadsOffInArea(
 			leftBottom->x, 
 			leftBottom->y, 
@@ -422,6 +462,20 @@ extern "C"
 			rightTop->z, 
 			true);
 
+		// For debugging
+		// Well this displays a message, why doesn't it turn off the paths for these roads?
+		std::string roadsMsg = "Ped roads have been switched off for: "
+		+ std::string("leftBottomX: ") + std::to_string(leftBottom->x) 
+		+ " leftBottomY: " + std::to_string(leftBottom->y)
+		+ " leftBottomZ: " + std::to_string(leftBottom->z)
+		
+		+ " rightTopX: " + std::to_string(rightTop->x)
+		+ " rightTopY: " + std::to_string(rightTop->y)
+		+ " rightTopZ: " + std::to_string(rightTop->z);
+
+		// LogFunctions::LogInfo(Defines::luaLogFile, roadsMsg);
+		//
+
 		return 0;
 	}
 
@@ -467,7 +521,7 @@ extern "C"
 	 */
 	int createObjectLua(lua_State *L)
 	{
-		return 1;
+		return 0;
 	}
 
 	/**
@@ -658,6 +712,9 @@ extern "C"
 
 	//-------------------
 	// Zone info
+	// Well I started working on these, although I'm not sure of the ScriptParams values for some parameters.
+	// I'll figure it out.
+	// Added these in v1.2.14-4a, they are still a work in progress.
 	//-------------------
 
 	/**
@@ -669,13 +726,157 @@ extern "C"
 	 * https://library.sannybuilder.com/#/vc/script/extensions/default/015C
 	 * 
 	 * Usage (gxtKey is a zone gxt key, dayOrNight is either 0 or 1.):
-	 * world.set_ped_zone_info(gxtkey, dayOrNight, density, cuban, haitian, street, diaz, security, biker, player, golfer, gang9, cop)
+	 * world.set_zone_ped_info(gxtkey, dayOrNight, density, cuban, haitian, street, diaz, security, biker, player, golfer, gang9, cop)
+	 */
+	int setZonePedInfoLua(lua_State *L)
+	{
+		const char *gxtLabel = lua_tostring(L, 1);
+
+		auto dayOrNight = luaL_checkinteger(L, 2);
+		auto density = luaL_checkinteger(L, 3);
+		auto cuban = luaL_checkinteger(L, 4);
+		
+		// 
+		auto haitian = luaL_checkinteger(L, 5);
+		auto street = luaL_checkinteger(L, 6);
+		//
+
+		bool diaz = luaL_checkinteger(L, 7);
+
+		auto security = luaL_checkinteger(L, 8);
+		auto biker = luaL_checkinteger(L, 9);
+		auto player = luaL_checkinteger(L, 10);
+		auto golfer = luaL_checkinteger(L, 11);
+		auto gang9 = luaL_checkinteger(L, 12);
+		auto cop = luaL_checkinteger(L, 13);
+		//
+
+		// TODO Test this, I'm not sure if it'll work.
+		int16 zone = CTheZones::FindZoneByLabelAndReturnIndex(const_cast<char*>(gxtLabel), ZONE_INFO);
+		if (zone < 0) {
+			return luaL_error(L, "world.set_zone_ped_info could not find zone!");
+		}
+
+		CTheZones::SetZonePedInfo(zone, 
+		dayOrNight, density, cuban, haitian, street, diaz, security,
+		biker, player, golfer, gang9, cop
+		);
+
+		return 0;
+	}
+
+	/**
+	 * Set the vehicle zone info.
+	 * 
+	 * TODO Fix this.
+	 * 
+	 * SET_ZONE_CAR_INFO
+	 * https://library.sannybuilder.com/#/vc/script/extensions/default/0152
+	 * 
+	 * Usage:
+	 * world.set_zone_car_info(gxtkey, dayOrNight, density, cuban, haitian, street, diaz, security, biker, player, golfer, gang9, cop)
+	 */
+	int setZoneCarInfoLua(lua_State *L)
+	{
+		int16 gangDensities[NUM_GANGS] = { 0 };
+		int i;
+
+		const char *gxtLabel = lua_tostring(L, 1);
+
+
+		// Set the gang densities.
+		// TODO Implement this part.
+		// for (i = 0; i < NUM_GANGS; i++)
+			// break;
+			// gangDensities[i] = ScriptParams[i + 2];
+
+		int16 zone = CTheZones::FindZoneByLabelAndReturnIndex(const_cast<char*>(gxtLabel), ZONE_INFO);
+
+		for (int i = 0; i < NUM_GANGS; i++) {
+			if (gangDensities[i] != 0 && CGangs::GetGangInfo(i)->m_nVehicleMI == -1)
+				debug("world.set_zone_car_info - Gang %d car ratio should be 0 in %s zone\n", i + 1, const_cast<char*>(gxtLabel));
+		}
+
+		if (zone < 0) {
+			return luaL_error(L, "world.set_zone_car_info could not find zone!");
+		}
+
+		// TODO Figure this out.
+		// while (zone >= 0) {
+		// 	CTheZones::SetZoneCarInfo(zone, ScriptParams[0], ScriptParams[1], ScriptParams[11], gangDensities);
+		// 	zone = CTheZones::FindNextZoneByLabelAndReturnIndex(label, ZONE_INFO);
+		// }
+
+		return 0;
+	}
+
+	/**
+	 * Set the civilian car zone info.
+	 * For setting up which ped vehicle classes spawn where.
+	 * 
+	 * TODO Fix this.
+	 * 
+	 * SET_ZONE_CIVILIAN_CAR_INFO
+	 * https://library.sannybuilder.com/#/vc/script/extensions/default/04EC
+	 * 
+	 * Usage:
+	 * world.set_zone_civilian_car_info(gxtkey, dayOrNight, normal, poor, rich, exec, worker, big, taxi, moped, motorbike, leisureBoat, workerBoat)
+	 */
+	int setZoneCivilianCarInfoLua(lua_State *L)
+	{
+		int16 carDensities[CCarCtrl::NUM_CAR_CLASSES] = { 0 };
+		int16 boatDensities[CCarCtrl::NUM_BOAT_CLASSES] = { 0 };
+		int i;
+
+		const char *gxtLabel = lua_tostring(L, 1);
+		auto dayOrNight = luaL_checkinteger(L, 2);
+
+		// TODO Figure this part out.
+		// for (i = 0; i < CCarCtrl::NUM_CAR_CLASSES; i++)
+			// carDensities[i] = ScriptParams[i + 1];
+
+		// for (i = 0; i < CCarCtrl::NUM_BOAT_CLASSES; i++)
+			// boatDensities[i] = ScriptParams[i + 1 + CCarCtrl::NUM_CAR_CLASSES];
+		//
+
+		int16 zone = CTheZones::FindZoneByLabelAndReturnIndex(const_cast<char*>(gxtLabel), ZONE_INFO);
+
+		if (zone < 0) {
+			return luaL_error(L, "world.set_zone_civilian_car_info could not find zone!");
+		}
+
+		while (zone >= 0) {
+			CTheZones::SetZoneCivilianCarInfo(zone, dayOrNight, carDensities, boatDensities);
+			zone = CTheZones::FindNextZoneByLabelAndReturnIndex(const_cast<char*>(gxtLabel), ZONE_INFO);
+		}
+
+		return 0;
+	}
+
+	/**
+	 * 
+	 * SET_ZONE_GROUP
+	 * https://library.sannybuilder.com/#/vc/script/extensions/default/0324
+	 * 
+	 * Usage:
+	 * world.set_zone_group(gxtkey, dayOrNight, pedGroup)
 	 */
-	int setPedZoneInfoLua(lua_State *L)
+	int setZoneGroupLua(lua_State *L)
 	{
-		// int density = 
-		// CTheZones::SetZonePedInfo()
+		const char *gxtLabel = lua_tostring(L, 1);
+		auto dayOrNight = luaL_checkinteger(L, 2);
+		auto pedGroup = luaL_checkinteger(L, 3);
+
+		int zone_id = CTheZones::FindZoneByLabelAndReturnIndex(const_cast<char*>(gxtLabel), ZONE_INFO);
 
+		if (zone_id < 0) {
+			return luaL_error(L, "world.set_zone_group: Couldn't find zone");
+		}
+
+		while (zone_id >= 0) {
+			CTheZones::SetPedGroup(zone_id, dayOrNight, pedGroup);
+			zone_id = CTheZones::FindNextZoneByLabelAndReturnIndex(const_cast<char*>(gxtLabel), ZONE_INFO);
+		}
 		return 0;
 	}
 
@@ -800,6 +1001,18 @@ extern "C"
 		return 0;
 	}
 
+	/**
+	 * Get a random cop in the area.
+	 * TODO Implement this.
+	 * 
+	 * Usage (Will return a CPed of a random cop in the area if there is one):
+	 * world.get_random_cop_in_area()
+	 */
+	// int getRandomCopInAreaLua(lua_State *L)
+	// {
+	// 	return 1;
+	// }
+
 	//-------------------
 
 	/**
diff --git a/src/extras/lua/lua_world.h b/src/extras/lua/lua_world.h
index ed0330bf..5492e797 100644
--- a/src/extras/lua/lua_world.h
+++ b/src/extras/lua/lua_world.h
@@ -52,6 +52,12 @@ extern "C" {
 	extern int createCarGeneratorLua(lua_State *L);
 	extern int switchCarGeneratorLua(lua_State *L);
 
+	// Zones
+	extern int setZonePedInfoLua(lua_State *L);
+	extern int setZoneCarInfoLua(lua_State *L);
+	extern int setZoneCivilianCarInfoLua(lua_State *L);
+	extern int setZoneGroupLua(lua_State *L);
+
 
 #ifdef __cplusplus
 }
diff --git a/src/extras/menus/misc_menu.cpp b/src/extras/menus/misc_menu.cpp
index ce213ebe..cf327016 100644
--- a/src/extras/menus/misc_menu.cpp
+++ b/src/extras/menus/misc_menu.cpp
@@ -684,6 +684,7 @@ void MiscMenu::MiscTab()
 		//
 	}
 
+	ImGui::Separator();
 
 	//-----------------------------------
 	// KCNet ReVC Lua
diff --git a/src/extras/utils/defines.cpp b/src/extras/utils/defines.cpp
index 932f6afc..410e743f 100644
--- a/src/extras/utils/defines.cpp
+++ b/src/extras/utils/defines.cpp
@@ -89,7 +89,7 @@ uint8_t Defines::reVCVersion = KCNET_REVC_VER_1_2_12;
 // TODO Make this grab from my latest internal gitea releases or something.
 // std::string Defines::programVersion = "1.1";
 #if defined(GTAVC)
-std::string Defines::programVersion = "1.2.14-3a";
+std::string Defines::programVersion = "1.2.14-4a";
 
 // Set the freeroam game name
 #ifdef DISABLE_GAME_SCRIPTS
diff --git a/src/extras/utils/freeroam_util.cpp b/src/extras/utils/freeroam_util.cpp
index 39fdf2e5..c5dfd433 100644
--- a/src/extras/utils/freeroam_util.cpp
+++ b/src/extras/utils/freeroam_util.cpp
@@ -10,10 +10,12 @@
 #include "vehicle_functions.h"
 
 // ReVC
+#include "Camera.h"
 #include "CarGen.h"
 #include "Clock.h"
 #include "Entity.h"
 #include "Frontend.h"
+#include "Pad.h"
 #include "Placeable.h"
 #include "PlayerInfo.h"
 #include "PlayerPed.h"
@@ -267,8 +269,11 @@ FreeroamUtil::SaveFreeroamJson()
 		//----------------
 		// Added in 1.2.14-3a
 		// For saving car generators on the map.
-		// TODO Test this, I'm not sure if it will work for the car generators.
+		// This seems to work, although it's not very useful for a freeroam.
+		// Currently, this is disabled and I probably won't use it anyways.
+		// It was mostly here for testing.
 
+#if 0
 		// Save the car generators.
 		// if(CTheCarGenerators::NumOfCarGenerators >= 0)
 		// {
@@ -280,10 +285,10 @@ FreeroamUtil::SaveFreeroamJson()
 		save["car_generators"]["process_counter"] = CTheCarGenerators::ProcessCounter;
 		save["car_generators"]["generate_if_player_is_close_counter"] = CTheCarGenerators::GenerateEvenIfPlayerIsCloseCounter;
 
-		
 
 		// TODO Figure this out, I need to save the status of these also.
 		// TODO Print the array list into JSON.
+
 		for (int i = 0; i < NUM_CARGENS; i++)
 		{
 			// json carGenArray = json::array(CTheCarGenerators::CarGeneratorArray[i].m_bForceSpawn);
@@ -294,9 +299,59 @@ FreeroamUtil::SaveFreeroamJson()
 			// save["car_generators"]["info"][i] = CTheCarGenerators::CarGeneratorArray[i];
 
 		}
-			// WriteSaveBuf(buffer, CarGeneratorArray[i]);
+#endif // DISABLED
+		
 
+		// TODO Add saving the below
+		// Most of the below is what the original game was saving, there is more but this is in GenericGameStorage.cpp.
+		// It is in the 'GenericSave' function.
+		// A lot of the original saving and data I'm never going to use.
+		// 
+
+		CPad *currentPad = CPad::GetPad(0);
+		// Save current pad mode
+		// currentPad->Mode
+
+
+		// These could be added, not sure if they are needed though.
+		// Save the camera position:
+		// TheCamera.GetPosition().x;
+		// TheCamera.GetPosition().y;
+		// TheCamera.GetPosition().z;
+
+		// Save the rest of the clock and timer values
+		// CClock::ms_nMillisecondsPerGameMinute
+		// CClock::ms_nLastClockTick
+		// CTimer::m_snTimeInMilliseconds
+		// CTimer::ms_fTimeScale
+		// CTimer::ms_fTimeStep
+		// CTimer::ms_fTimeStepNonClipped
+		// CTimer::m_FrameCounter
+		// CTimeStep::ms_fTimeStep
+		// CTimeStep::ms_fFramesPerUpdate
+		// CTimeStep::ms_fTimeScale
 		//
+
+		// Save the cameras zoom indicators
+		// TheCamera.CarZoomIndicator
+		// TheCamera.PedZoomIndicator
+		// 
+
+		// Inverting look for the controller
+		// CPad::bInvertLook4Pad
+		//
+
+		// Time cycle extra colors
+		// CTimeCycle::m_ExtraColour
+		// CTimeCycle::m_bExtraColourOn
+		// CTimeCycle::m_ExtraColourInter
+
+		// TODO Figure out what these radio functions are saving
+		// It seems to be the last slot the radio was on.
+		// PopulateRadioStationPositionList
+		// WriteDataToBufferPointer(buf, RadioStationPosition);
+
+		// End saving the JSON file
 		//----------------
 
 		/*
diff --git a/src/save/GenericGameStorage.cpp b/src/save/GenericGameStorage.cpp
index a7d90ac2..8787eca2 100644
--- a/src/save/GenericGameStorage.cpp
+++ b/src/save/GenericGameStorage.cpp
@@ -46,6 +46,10 @@
 #include "defines.h"
 #endif // CUSTOM_CODE
 
+#ifdef DISABLE_GAME_SCRIPTS
+#include "freeroam_util.h"
+#endif // DISABLE_GAME_SCRIPTS - Load/save custom json format for saves.
+
 
 #define BLOCK_COUNT 22
 #define SIZE_OF_SIMPLEVARS 0xE4
@@ -75,6 +79,9 @@ uint32 TimeToStayFadedBeforeFadeOut = 1750;
 
 int32 RadioStationPosition[NUM_RADIOS];
 
+/**
+ * Setup the radio station list for the saves.
+ */
 void
 InitRadioStationPositionList()
 {
@@ -82,12 +89,18 @@ InitRadioStationPositionList()
 		RadioStationPosition[i] = -1;
 }
 
+/**
+ * Get the last radio position for the saves.
+ */
 int32
 GetSavedRadioStationPosition(int32 station)
 {
 	return RadioStationPosition[station];
 }
 
+/**
+ * Populate the radio position list the saves.
+ */
 void
 PopulateRadioStationPositionList()
 {
@@ -142,33 +155,12 @@ GenericSave(int file)
 	// TODO Fix this to disable loading of the game if enabled in the ini.
 	// Early exit out of this, game scripts are disabled and this will crash.
 
-	// TODO Make this save just a couple of values and the stats to a freeroam-game-save.json or something.
-	
-	// Possibly save these values
-	// WriteSaveDataBlock(CTheCarGenerators::SaveAllCarGenerators, "AllCarGeneratorsSize");
-	//	WriteSaveDataBlock(CRadar::SaveAllRadarBlips, "RadarBlipsBufferSize");
-	// 	WriteSaveDataBlock(CRestart::SaveAllRestartPoints, "RestartPointsBufferSize");
-	// 	WriteSaveDataBlock(CPickups::Save, "PickUpsSize");
-
-	// This one will be complicated if i decide to implement garages in the freeroam
-	// WriteSaveDataBlock(CGarages::Save, "GaragesSize");
-
-	// This saves the players info, maybe I should just do something like this to write to my save file.
-	// TODO Save this to a json file or something.
-	// 	WriteSaveDataBlock(CWorld::Players[CWorld::PlayerInFocus].SavePlayerInfo, "PlayerInfoSize");
-
-	// Save the player stats
-	// WriteSaveDataBlock(CStats::SaveStats, "StatsSize");
-	
+	// TODO Test this later
+	// I will need to implement a save pickup that can respawn when picked up.
+	// FreeroamUtil::SaveFreeroamJson();
 	//
 
-	// Save the weather
-	// WriteDataToBufferPointer(buf, CWeather::OldWeatherType);
-	// WriteDataToBufferPointer(buf, CWeather::NewWeatherType);
-	// WriteDataToBufferPointer(buf, CWeather::ForcedWeatherType);
-	// WriteDataToBufferPointer(buf, CWeather::WeatherTypeInList);
-	// 
-	
+
 	return true;
 
 	// Always disable saving code below, to prevent it from running.
@@ -351,6 +343,9 @@ GenericLoad()
 	// This breaks the new game if set here, I cannot press F5 to restart the script or press new game without crashing.
 	// LuaTest::GetInstance().SetupFreeroamScript();
 	
+	// TODO Test this later
+	// FreeroamUtil::LoadFreeroamJson();
+	//
 
 	return true;
 
@@ -558,6 +553,10 @@ CloseFile(int32 file)
 	return CFileMgr::CloseFile(file) == 0;
 }
 
+/**
+ * Load the collision and everything
+ * TODO Make my game scripts do some of this.
+ */
 void
 DoGameSpecificStuffAfterSucessLoad()
 {
@@ -571,11 +570,21 @@ DoGameSpecificStuffAfterSucessLoad()
 	CTheScripts::Process();
 }
 
+/**
+ * Check if the save data is valid.
+ * 
+ * TODO Add support to check if my json file has the correct version and format.
+ */
 bool
 CheckSlotDataValid(int32 slot)
 {
 	PcSaveHelper.nErrorCode = SAVESTATUS_SUCCESSFUL;
+
+#ifdef DISABLE_GAME_SCRIPTS
 	if (CheckDataNotCorrupt(slot, LoadFileName)) {
+#else
+	if (CheckDataNotCorrupt(slot, LoadFileName)) {
+#endif // DISABLE_GAME_SCRIPTS
 		CStreaming::DeleteAllRwObjects();
 		return true;
 	}
@@ -601,6 +610,9 @@ CopySizeAndPreparePointer(uint8 *&buf, uint8 *&postbuf, uint8 *&postbuf2, uint32
 	postbuf = postbuf2;
 }
 
+/**
+ * TODO Look into using this before I save my game with my json format.
+ */
 void
 DoGameSpecificStuffBeforeSave()
 {
@@ -649,6 +661,8 @@ CheckDataNotCorrupt(int32 slot, char *name)
 	if (file == 0)
 		return false;
 	strcpy(name, filename);
+
+
 	while (SIZE_OF_ONE_GAME_IN_BYTES - sizeof(uint32) > bytes_processed && blocknum < 40) {
 		int32 blocksize;
 		if (!ReadDataFromFile(file, (uint8*)&blocksize, sizeof(blocksize))) {
@@ -692,6 +706,9 @@ CheckDataNotCorrupt(int32 slot, char *name)
 	return false;
 }
 
+/**
+ * TODO Look into this function later for my save system.
+ */
 bool
 RestoreForStartLoad()
 {
-- 
2.43.0

